### ISSUE TITLE
Error tolerance control

### ISSUE THREAD

**Issue by m13253** [m13253/lvdo#4]
Is it necessary to have an error tolerance control mechanism?

I have to sides of view:

Positive:
- Data should be transported reliably and integrity be kept
- Corrupted data causes noise on decoding or even decode failure

Negative:
- Error tolerance requires more data rate (50% in estimate), which is insufficient
- Integrity can be built on top of transport layer, following the example of OSI model
- Lossy media file does not require much data integrity
- Different algorithms can be chosen, e.g. Hamming Code, to keep data integrity, of which the decision depends on different types of file
- Transport error, alongside with rounding error, can be prevent by choosing moderated or conservative parameters

Maybe there should be another Transmission Control Layer built on top of LVDO Transport Layer, with advanced data rate control, framerate control, Youku watermark bypass solution, and H.264 motion estimate bypass solution.

**Comment by m13253** [comment]
可以看一下类似 WinRAR “自我修复”的功能。
现代无线电通信也在使用这样的功能。

那就是，100% 的数据包，成功接收到其中任意 90% 的数据包（这个阈值可控），就可以修复损坏的数据，并且解码出全部的数据。

如果你感兴趣可以试着实现它。

**Comment by Skylion007** [comment]
I actually did implement this feature in Python that links to my LVDO Windows fork. The algorithm I used is Reed-Solomon encoding and it worked quite well. I believe that is what most radio operators use. While I don't understand all the math behind it, you can find several implementations of it online.

**Comment by felixonmars** [comment]
IIRC Reed-Solomon was used in the Parchive project, and libpar2 should be available in most distro.

**Comment by Skylion007** [comment]
Yeah I actually used Par archives before switching over to Reed-Solomon. There are tons of libraries for the algorithm written in C as well such as https://github.com/catid/longhair . Googling will yield a lot of results. The main issue is just integrating it into the repo.

**Comment by m13253** [comment]
> I actually did implement this feature in Python that links to my LVDO Windows fork.

I am very happy to see friends from all over the world who pay attention to my experiment.
LVDO was originally written as a hobby project, to state against the increasingly strict censorship of online videos conducted by Chinese government. (By the time this repo was started, a Chinese video site named [Sina Video was shut down](http://www.marbridgeconsulting.com/marbridgedaily/2014-04-24/article/74440/sinas_online_publication_video_licenses_revoked_for_hosting_illicit_content), just because the web master did not delete a porn video in time before [MOC](https://en.wikipedia.org/wiki/Ministry_of_Culture_of_the_People%27s_Republic_of_China) found it)
It is [unsafe for me to write these code](https://www.eff.org/deeplinks/2015/08/speech-enables-speech-china-takes-aim-its-coders), also I do not have enough time to work on it. So I will appreciate if anyone could continue this experiment.
